Molecule rotation

pjanik opened the pull request:
This PR adds a possibility to rotate a molecule.
To do it, you need to press Alt / Opt key and drag one atom of the molecule.

There's also a new MD2D option called `onAtomDrag`. It's default value is `translate`, but author can set it to `rotate` and then the default drag behavior would be rotation. The non-default behavior can be always accessed using Alt / Opt key.

scytacki commented on src/lab/models/md2d/views/atoms-interactions.js:
```diff
@@ -222,61 +252,57 @@ define(function(require) {
+      var translate = mode === TRANSLATE;
       // Fast path, no dragging at all if model is running and atom isn't draggable.
-      if ( ! isAtomDraggable(atom) ) return;
+      if (translate && !isAtomDraggable(atom)) return;
```
It is somewhat confusing that the 'fast path' for translate is here, but the 'fast path' for rotation happens earlier in `mouseDown`. I'm not sure if it would look any better but perhaps this fastpath check could be moved to mouseDown.

pjanik replied:
Right, I've done it.

scytacki commented on src/lab/models/md2d/views/atoms-interactions.js:
```diff
@@ -222,61 +252,57 @@ define(function(require) {
-        // Custom drag handler.
-        if (modelView.dragHandler.atom) {
-          modelView.dragHandler.atom(x, y, atom, i);
-        }
```
Do any of the custom drag handlers try to set the cursor themselves? If so I think these changes will break that behavior, since setCursor is now called after the dragHandler.  Would it make since to set the cursor before translateMolecule

pjanik replied:
I don't think so, as there's no way to do it using scripting API AFAIR. But I moved custom drag handler below anyway, so it's theoretically possible in the future, it won't hurt.

scytacki commented on src/lab/models/md2d/views/atoms-interactions.js:
```diff
@@ -285,27 +311,88 @@ define(function(require) {
+      if (bbox.top + y > modelHeight) y = modelHeight - bbox.top;
+
+      if (model.isStopped()) {
+        setAtomPosition(atom.idx, x, y, false, true);
```
Does setAtomPosition actually change all of the atoms in the molecule? I see in rotate molecule you have to iterate over all of the atoms but here you don't.

pjanik replied:
It depends on arguments that we pass to `.setAtomProperties`. One of them is called "moveMolecule" and if it's true, it moves all of the atoms.

I cleanup that a bit and provided separate functions that either move atom or molecule, so it's more obvious when you read this code.

scytacki commented:
Looks good. I just had a few inline comments and clarifications.

pjanik commented:
Thanks for good feedback. I've addressed all the comments and also made two new changes:
- rotation respects `atomIsDraggable` setting, so author can control which atoms / molecules can be rotated
- custom dragHandler is executed when we rotate molecule

Basically, I've assumed that dragging is just a generic action and it can start either translation or rotation. So all the dragging-related settings should affect both behaviors.

scytacki commented:
👍
